tsl/src/fdw/data_node_scan_plan.c

```diff
@@ -741,6 +755,691 @@ push_down_group_bys(PlannerInfo *root, RelOptInfo *hyper_rel, Hyperspace *hs,
+	 * (see timescaledb_set_join_pathlist_hook).
+	 */
+	TimescaleDBPrivate *outerrel_private = outerrel->fdw_private;
+	Assert(outerrel_private != NULL);
+	Assert(outerrel_private->fdw_relation_info != NULL);
+
+	/* Get the hypertable from the outer relation. */
+	RangeTblEntry *rte_outer = planner_rt_fetch(outerrel->relid, root);
+
+	/* Test that the fetched outer relation is a distributed hypertable. If not,
+	 * the pushdown is not possible. This should be already tested by the previous
+	 * checks on 'outerrel'. */
+	Assert(rte_outer != NULL);
+	Assert(is_distributed_hypertable(rte_outer->relid));
+#endif
+
+	/* Test if inner table has a range table. */
+	RangeTblEntry *rte_inner = planner_rt_fetch(innerrel->relid, root);
+	if (rte_inner == NULL)
+		return;
+
+	/* Get current partitioning of the outerrel. */
+	hyper_table_rels = outerrel->part_rels;
+	nhyper_table_rels = outerrel->nparts;
+
+	Assert(nhyper_table_rels > 0);
+	Assert(hyper_table_rels != NULL);
+
+	/*
+	 * Create unfinished PgFdwRelationInfo entry which is used to indicate
+	 * that the join relation is already considered, so that we won't waste
+	 * time in judging safety of join pushdown and adding the same paths again
+	 * if found safe. Once we know that this join can be pushed down, we fill
+	 * the entry.
+	 */
+	fpinfo =
+		fdw_relinfo_create(root, joinrel, InvalidOid, InvalidOid, TS_FDW_RELINFO_UNINITIALIZED);
+	Assert(fpinfo->type == TS_FDW_RELINFO_UNINITIALIZED);
+
+	/* attrs_used is only for base relations. */
+	fpinfo->attrs_used = NULL;
+	fpinfo->pushdown_safe = false;
+
+	/*
+	 * We need the FDW information to get retrieve the information about the
+	 * configured reference join tables. So, create the data structure for
+	 * the first server. The reference tables are the same for all servers.
+	 */
+	Oid server_oid = hyper_table_rels[0]->serverid;
+	fpinfo->server = GetForeignServer(server_oid);
+	apply_fdw_and_server_options(fpinfo);
+
+	if (!is_safe_to_pushdown_reftable_join(root,
+										   fpinfo->join_reference_tables,
+										   rte_inner,
+										   jointype))
+	{
+		/*
+		 * Reset fdw_private to allow further planner calls with different arguments
+		 * (e.g., swapped inner and outer relation) to replan the pushdown.
+		 */
+		pfree(joinrel->fdw_private);
+		joinrel->fdw_private = NULL;
+		return;
+	}
+
+	/*
+	 * Join pushdown only works if the data node rels are created in
+	 * data_node_scan_add_node_paths during scan planning.
+	 */
+	if (!ts_guc_enable_per_data_node_queries)
+	{
+		ereport(DEBUG1,
+				(errmsg("join on reference table is not considered to be pushed down because "
+						"'enable_per_data_node_queries' GUC is disabled")));
+
+		return;
+	}
+
+	/* The inner table can be a distributed hypertable or a plain table. Plain tables don't have
+	 * a TsFdwRelInfo at this point. So, it needs to be created.
+	 */
+	if (innerrel->fdw_private == NULL)
+		fdw_relinfo_create(root, innerrel, InvalidOid, InvalidOid, TS_FDW_RELINFO_REFJOIN_TABLE);
```

Comment:
Does it really need an fpinfo?

Also please note that it might be a table managed by another extension (e.g. `postgres_fdw`). This is not likely, but we have to be careful about accessing `fdw_private` in such cases. The other extensions are the reason why we don't touch the `fdw_private` field unless our `classify_relation` says it's our table.

Also the name is somewhat confusing, maybe `TS_FDW_RELINFO_REFERENCE_TABLE`?

Reply:
> Does it really need an fpinfo?

Oh, it probably does, at least we're storing the local/remote conditions there.

Reply:
At this point, the `innerrel` has to be marked as a reference table. So it has to be 'our' table. Depending on the used replication method, it could be a DHT or a plain PostgreSQL table. I am not sure that `classify_relation` would help much here to protect this table.

As suggested `TS_FDW_RELINFO_REFJOIN_TABLE` was renamed to `TS_FDW_RELINFO_REFERENCE_TABLE`.